subsys/net/ip/tcp.c

```diff
@@ -2639,124 +2666,205 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
+			} else {
+				if (len > 0) {
+					/* Send out a duplicate ACK */
+					tcp_out(conn, ACK);
+					verdict = NET_OK;
+				}
 			}
 		}
 		break;
 	case TCP_CLOSING:
-		if (th && FL(&fl, ==, ACK, th_seq(th) == conn->ack)) {
-			tcp_send_timer_cancel(conn);
-			next = TCP_TIME_WAIT;
-			verdict = NET_OK;
+		if (th) {
+			/*
+			 * Closing:
+			 * Our FIN has to be acknowledged
+			 * -   -> TCP_CLOSING
+			 * ACK -> TCP_TIME_WAIT
+			 */
+			int32_t new_len = tcp_compute_new_length(conn, th, len);
+			/* Count with -1 as the FIN flag has to be subtracted */
+			if ((new_len > -1) && (len > 0)) {
+				/* This should not happen here, as no data can be send after
+				 * the FIN flag has been send.
+				 */
+				NET_ERR("conn: %p, new bytes %u during CLOSING state "
+					"sending reset", conn, new_len);
+				net_stats_update_tcp_seg_drop(conn->iface);
+
+				next = tcp_enter_time_wait(conn);
+
+				tcp_out(conn, RST);
+				break;
+			}
+
+			if (FL(&fl, &, ACK, th_ack(th) == conn->seq)) {
+				NET_DBG("FIN acknowledged, going to TIME WAIT "
+					"state seq %u, ack %u"
+					, conn->seq, conn->ack);
+
+				next = tcp_enter_time_wait(conn);
+
+				verdict = NET_OK;
+			}
+
+			/*
+			 * There can also be data in the message, so compute with the length
+			 * of the packet to check with the ack
+			 */
+			if ((FL(&fl, &, FIN, (th_seq(th) + len) == conn->ack)) || (len > 0)) {
+				if (next == TCP_TIME_WAIT) {
+					/* Send out a duplicate ACK */
+					tcp_out(conn, ACK);
+				} else {
+					/* Send out a duplicate ACK, with the pending FIN
+					 * flag
+					 */
+					tcp_send_timer_cancel(conn);
+					tcp_out_ext(conn, FIN | ACK, NULL, conn->seq - 1);
+				}
+				verdict = NET_OK;
+			}
 		}
 		break;
 	case TCP_TIME_WAIT:
-		/* Acknowledge any FIN attempts, in case retransmission took
-		 * place.
-		 */
-		if (th && (FL(&fl, ==, (FIN | ACK), th_seq(th) + 1 == conn->ack) ||
-			   FL(&fl, ==, FIN, th_seq(th) + 1 == conn->ack))) {
-			tcp_out(conn, ACK);
-			verdict = NET_OK;
-		}
+		if (th) {
+			int32_t new_len = tcp_compute_new_length(conn, th, len);
+			/* Count with -1 as the FIN flag has to be subtracted */
+			if ((new_len > -1) && (len > 0)) {
+				/* This should not happen here, as no data can be send after
+				 * the FIN flag has been send.
+				 */
+				NET_ERR("conn: %p, new bytes %u during TIME-WAIT state "
+					"sending reset", conn, new_len);
+				net_stats_update_tcp_seg_drop(conn->iface);
 
-		k_work_reschedule_for_queue(
-			&tcp_work_q, &conn->timewait_timer,
-			K_MSEC(CONFIG_NET_TCP_TIME_WAIT_DELAY));
+				tcp_out(conn, RST);
+
+				k_work_reschedule_for_queue(
+					&tcp_work_q, &conn->timewait_timer,
+					K_MSEC(CONFIG_NET_TCP_TIME_WAIT_DELAY));
+			} else {
+				/* Acknowledge any FIN attempts, in case retransmission took
+				 * place.
+				 */
+				if ((FL(&fl, &, FIN, th_seq(th) + 1 == conn->ack)) || (len > 0)) {
+					tcp_out(conn, ACK);
+					verdict = NET_OK;
+
+					k_work_reschedule_for_queue(
```

Comment:
Is there a need to reset TIME_WAIT timer? It seems quite unsafe to me - a malfunctioning (or malicious) peer could keep sending FIN packets, effectively preventing us from releasing resources.